### PR TITLE
fix: override `eth_getProof` to enable withdrawals

### LINKED ISSUE #72: Withdrawal proof cache
Transferred over from https://github.com/paradigmxyz/alphanet/issues/135

# Motivation

To withdraw from an OP stack chain, a user needs to provide a proof on L1 that a withdrawal was included on-chain on L2. To do this, `eth_getProof` is called to prove a storage slot on the withdrawal contract against an output root.

On Reth, `eth_getProof` is really expensive due to no historical tries being stored. To prove at an old block, old state is reconstructed in-memory using changesets. The further back you go, the more expensive this operation is, which is why it is [restricted] in Reth.

We *could* increase this maximum, but it would likely be too impractical to run a node with the resources necessary to actually compute historical proofs on a 1s blocktime chain with activity, gigagas or not.

To solve this, we are going to build an ExEx.

# Proposed solution

We will build an ExEx that computes withdrawal proofs and caches them in a separate database. The `eth_getProof` RPC method is replaced with one that first checks this cache, and thereafter computes a historical proof as usual with the normal constraints on how far back you can go.

The ExEx will look for any new withdrawals and store these. Then, every time an output root is posted on L1, non-proven withdrawals are proven and the resulting proof is stored in the cache.

### PR DESCRIPTION
**Pull request by joshieDo**
closes https://github.com/ithacaxyz/odyssey/issues/84 and closes #72 

Overrides `eth_getProof` as to enable withdrawals
```rust
//! - `eth_getProof` will _ONLY_ return the storage proofs _WITHOUT_ an account proof _IF_ targeting
//!   the withdrawal contract. Otherwise, it fallbacks to default behaviour.
```

### CODE REVIEW

File: crates/node/src/rpc.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        &self,
+        address: Address,
+        keys: Vec<JsonStorageKey>,
+        block_number: Option<BlockId>,
+    ) -> RpcResult<EIP1186AccountProofResponse> {
+        trace!(target: "rpc::eth", ?address, ?keys, ?block_number, "Serving eth_getProof");
+
+        // If we are targeting the withdrawal contract, then we only need to provide the storage
+        // proofs for withdrawal.
+        if address == WITHDRAWAL_CONTRACT {
```
Comment (mattsse) [inline crates/node/src/rpc.rs:72]: is this something we want to bake into op-reth as well?

File: crates/node/src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+use reth_rpc_eth_api::{
+    helpers::{EthState, FullEthApi},
+    FromEthApiError,
+};
+use reth_rpc_eth_types::EthApiError;
+use reth_rpc_types_compat::proof::from_primitive_account_proof;
+use reth_trie_common::AccountProof;
+use tracing::trace;
+
+const WITHDRAWAL_CONTRACT: alloy_primitives::Address =
```
Comment (onbjerg) [inline crates/node/src/rpc.rs:26]: nit: docs on this would be nice

**Review by onbjerg — COMMENTED** [review]
I'd also like an e2e test, one for proving the withdrawal contract, one for proving something that isn't the withdrawal contract (to ensure the fallback works through refactors e.g.)

can take a look at our current e2e tests

**Review by onbjerg — APPROVED** [review]
looks good to me, great work bud

**Pull request merged**